Add support for PSL onehot/onehot0 functions

tmeissner opened the pull request:
This is an attempt to add support for PSL function `onehot()` and `onehot0()`for synthesis, at least.
**Ready for review**

tgingold commented:
You need to declare the identifiers in std_names.ad[sb]

tmeissner commented:
Hmm, I am not able to decrypt these errors during some python tests. In the LanguageServer for example.

eine commented:
FTR, I just executed `./scripts/update_py_bindings.sh`, committed the changes and pushed on top of this.

We need to fix CI because job pyGHDL should have failed when @tmeissner pushed. Instead of `git diff --quiet || git status`, we need something that fails.

tmeissner commented:
> FTR, I just executed `./scripts/update_py_bindings.sh`, committed the changes and pushed on top of this.

Thanks. Seems to build successfully now. So, I can forward with synthesis related changes 🙂

tmeissner commented:
I think I've done all necessary changes except implementation of the one-hot algorithm itself. I her to dig deeper in the synthesis code base before I have the confidence to do that. At the moment I'm a little bit confused of the different types. And how to select a specific bit of a vector. Or how to get a constant vector of zero for example 🧐

tgingold commented:
The most useful builders are in netlists-folds and netlists-builders.

So to extract a bit of a vector, you can use build2_extract.
To create a constant vector, use build2_const_uns.

tmeissner commented:
I have something working now. Sadly I have problems with git (not the first time, I hate it), which have to be solved before I can push.

And, the code don't works with scalar types like integer. I get a crash with assertion in the `Vec_Length()` function.
```
raised CONSTRAINT_ERROR : synth-objtypes.adb:533 discriminant check failed
```
I assume, the vector has to be converted before I can use it with these function.

tgingold commented:
Vec_Len applies only to vectors.  Just use `Val.Typ.W` which is the width of the net.  This works for onehot, as the value is either a vector of bit or a number (which is represented by a vector of bit).

tmeissner commented:
I've added test cases and updated the documentation. Ready for review.
@tgingold I'm not absolutely sure about behavior of the two functions with integers < 0. I assume that these are encoded in two-complement, but that's only a guess. And which encoding they have in the during synthesis.

cmarqu commented on src/synth/synth-expr.adb:
```diff
@@ -2057,6 +2057,109 @@ package body Synth.Expr is
+      DffCurr := Get_Net (Ctxt, Expr);
+      Set_Location (DffCurr, Call);
+
+      -- Create a constant vector of 0 to comparation
```
```suggestion
      -- Create a constant vector of 0 for comparing
```

tgingold commented:
Is onehot(x) = onehot0(not x) ?  If so, just write the onehot once.
If not, explain!

tmeissner commented:
> Is onehot(x) = onehot0(not x) ? If so, just write the onehot once. If not, explain!

`onehot0(x)` = `onehot(x) or 0`

tgingold commented:
Ok, I should have looked at the standard instead of relying on my memory!

But, my comment still partially apply: factorize the code that compute onehot (for a net).

You don't need to set the location for Get_Net of a constant.

Nice work!

tmeissner commented:
> Ok, I should have looked at the standard instead of relying on my memory! But, my comment still partially apply: factorize the code that compute onehot (for a net). You don't need to set the location for Get_Net of a constant. Nice work!

Okay, I already thought about to factorize the one-hot logic in an own function. But I didn't know if that id desirable.

tmeissner commented:
As with the used algorithm, `onehot0` is simpler than `onehot`, I refactored `onehot0` in an own function. I don't know if function overloading is welcome ... I now have something like this:

```ada
   function Synth_Psl_Onehot0 (Ctxt : Context_Acc; DffCurr : Net; Call : Node;
                               Vlen : Uns32)
                               return Net
   is
      DffZero   : Net;
      DffOne    : Net;
      DffSub    : Net;
      DffSubAnd : Net;
      Res       : Net;
   begin
      -- Create a constant vector of 0 to comparation
      DffZero := Build2_Const_Uns(Ctxt, 0,  Vlen);

      -- Get vector of value 1 for subtraction
      DffOne := Build2_Const_Uns(Ctxt, 1,  Vlen);

      -- Subtraction -> v - 1
      DffSub := Build_Dyadic (Ctxt, Id_Sub, DffCurr, DffOne);
      Set_Location (DffSub, Call);

      -- Binary And -> v & (v - 1)
      DffSubAnd := Build_Dyadic (Ctxt, Id_And, DffCurr, DffSub);
      Set_Location (DffSubAnd, Call);

      -- Compare with 0 -> (v & (v - 1)) == 0
      Res := Build_Compare (Ctxt, Id_Eq, DffSubAnd, DffZero);
      Set_Location (Res, Call);
      
      return Res;
   end Synth_Psl_Onehot0;

   function Synth_Psl_Onehot (Syn_Inst : Synth_Instance_Acc; Call : Node)
                              return Valtyp
   is
      Ctxt             : constant Context_Acc := Get_Build (Syn_Inst);
      Expr             : Valtyp;
      DffCurr          : Net;
      DffCurrIsZero    : Net;
      DffCurrIsNotZero : Net;
      DffOneHot0       : Net;
      Res              : Net;
      Vlen             : Uns32;
   begin
      -- Get parameter & its length
      Expr := Synth_Expression (Syn_Inst, Get_Expression (Call));
      Vlen := Expr.Typ.W;

      -- First get net of parameter
      DffCurr := Get_Net (Ctxt, Expr);
      Set_Location (DffCurr, Call);

      -- Compare parameter with 0 -> v = 0
      DffCurrIsZero := Build_Compare (Ctxt, Id_Eq, DffCurr,
                                      Build2_Const_Uns(Ctxt, 0,  Vlen));
      Set_Location (DffCurrIsZero, Call);

      -- Negate Compare result -> !(v = 0)
      DffCurrIsNotZero := Build_Monadic(Ctxt, Id_Not, DffCurrIsZero);
      Set_Location (DffCurrIsNotZero, Call);

      -- Synth onehot0
      DffOneHot0 := Synth_Psl_Onehot0 (Ctxt, DffCurr, Call, Vlen);
      Set_Location (DffOneHot0, Call);

      -- Final Binary And -> !(v = 0) & ((v & (v - 1)) == 0)
      Res := Build_Dyadic (Ctxt, Id_And, DffOneHot0, DffCurrIsNotZero);
      Set_Location (Res, Call);

      return Create_Value_Net (Res, Boolean_Type);
   end Synth_Psl_Onehot;

   function Synth_Psl_Onehot0 (Syn_Inst : Synth_Instance_Acc; Call : Node)
                               return Valtyp
   is
      Ctxt    : constant Context_Acc := Get_Build (Syn_Inst);
      Expr    : Valtyp;
      Vlen    : Uns32;
      DffCurr : Net;
      Res     : Net;
   begin
      -- Get parameter & its length
      Expr := Synth_Expression (Syn_Inst, Get_Expression (Call));
      Vlen := Expr.Typ.W;

      -- First get net of parameter
      DffCurr := Get_Net (Ctxt, Expr);
      Set_Location (DffCurr, Call);

      Res := Synth_Psl_Onehot0 (Ctxt, DffCurr, Call, Vlen);

      return Create_Value_Net (Res, Boolean_Type);
   end Synth_Psl_Onehot0;
```

tmeissner commented:
Okay. I will try to optimize the logic.

tmeissner commented:
Feel free to take my commits & refactor or optimize further. I'm not a SW guy and I'm not that experienced with development of big SW-projects 😄

tgingold commented:
Ok.  Thank you for this work!
